plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py

```diff
@@ -259,76 +274,122 @@ def check_link_down_condition(self, port_obj, port_name, ports_counters):
+                return None
+            for (interval, threshold) in self.ber_intervals:
+                symbol_ber_rate = self.calc_ber_rates(port_name, port_obj.active_speed, port_obj.port_width, interval)
+                if symbol_ber_rate and symbol_ber_rate > threshold:
+                    return Issue(port_name, Constants.ISSUE_BER)
+        return None
+
     def read_next_set_of_high_ber_or_pdr_ports(self, endpoint_port):
+        """
+        Read the next set of ports and check if they have high BER, PDR, temperature or link downed issues
+        """
         issues = {}
         ports_counters = self.ufm_client.get_telemetry(endpoint_port, Constants.PDR_DYNAMIC_NAME,self.test_mode)
         if ports_counters is None:
-            # implement health check for the telemetry instance
             self.logger.error("Couldn't retrieve telemetry data")
-            return issues
+            raise DynamicTelemetryUnresponsive
         for index, row in ports_counters.iterrows():
             port_name = f"{row.get('port_guid', '').split('x')[-1]}_{row.get('port_num', '')}"
             if self.test_mode:
-                # I want to add them once we get our first telemetry.
+                # Adding the port data upon fetching the first telemetry data
                 if not self.ports_data.get(port_name):
                     self.ports_data[port_name] = PortData(port_name)
             port_obj = self.ports_data.get(port_name)
             if not port_obj:
                 self.logger.warning("Port {0} not found in ports data".format(port_name))
                 continue
-            # from micro seconds to seconds.
+            # Converting from micro seconds to seconds.
             timestamp = row.get(Constants.TIMESTAMP) / 1000 / 1000
-            rcv_error = get_counter(Constants.RCV_ERRORS_COUNTER, row)
-            rcv_remote_phy_error = get_counter(Constants.RCV_REMOTE_PHY_ERROR_COUNTER, row)
-            errors = rcv_error + rcv_remote_phy_error
-            error_rate = self.get_rate_and_update(port_obj, Constants.ERRORS_COUNTER, errors, timestamp)
-            rcv_pkts = get_counter(Constants.RCV_PACKETS_COUNTER, row)
-            rcv_pkt_rate = self.get_rate_and_update(port_obj, Constants.RCV_PACKETS_COUNTER, rcv_pkts, timestamp)
-            cable_temp = get_counter(Constants.TEMP_COUNTER, row, default=None)
-            link_downed = get_counter(Constants.LNK_DOWNED_COUNTER, row)
-            link_downed_rate = self.get_rate_and_update(port_obj, Constants.LNK_DOWNED_COUNTER, link_downed, timestamp)
+            
+            pdr_issue = self.check_pdr_issue(port_obj, port_name, row, timestamp)
+            temp_issue = self.check_temp_issue(port_obj, port_name, row, timestamp)
+            link_downed_issue = self.check_link_down_issue(port_obj, port_name, row, timestamp, ports_counters)
+            ber_issue = self.check_ber_issue(port_obj, port_name, row, timestamp)
             port_obj.last_timestamp = timestamp
```

Comment:
because you only put one item and discrad the rest you can write the if between, and no call the other functions for example:
pdr_issue = self.check_pdr_issue(..)
if pdr_issue:
    issues[port_name] = pdr_issue
    return issues
temp_issue = ...
if temp_issue: ..

This change will reduce amount of checks that you are doing now

Reply:
good comment, fixed